### ISSUE TITLE
rendering error in Black Myth

### ISSUE THREAD

**Issue by jaskang** [veeenu/hudhook#203]
The same code can run in "Elden Ring" and work, but it will crash in "Black Myth".

I am using sample code： examples/demo_hook_dx12.rs

Error:
```
Render error: Error { code: HRESULT(0x887A0005), message: "The GPU device instance has been suspended. Use GetDeviceRemovedReason to determine the appropriate action." }
```

Elden Ring
![image](https://github.com/user-attachments/assets/aa85236f-b3ee-43d2-a85b-b030993b8e9e)

Black Myth Wukong
![image](https://github.com/user-attachments/assets/be9e7d3d-b01d-40e9-a1b8-e5b47644df95)

log:

```
2024-09-19T12:30:34.434333Z TRACE ThreadId(01) black_myth_map: src\lib.rs:79: DllMain()
2024-09-19T12:30:34.436849Z DEBUG ThreadId(02) hudhook::hooks: C:\hudhook-0.7.1\src\hooks\mod.rs:87: WNDCLASSEXW { cbSize: 80, style: WNDCLASS_STYLES(3), cbClsExtra: 0, cbWndExtra: 0, hInstance: HINSTANCE(5368709120), hIcon: HICON(0), hCursor: HCURSOR(0), hbrBackground: HBRUSH(0), lpszMenuName: PCWSTR(0x0), lpszClassName: PCWSTR(0x7ffe3ef1dcf8), hIconSm: HICON(0) }
2024-09-19T12:30:34.440773Z DEBUG ThreadId(02) hudhook::hooks: C:\hudhook-0.7.1\src\hooks\mod.rs:107: HWND(1380404)
2024-09-19T12:30:34.490453Z TRACE ThreadId(02) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:259: IDXGISwapChain::Present = 0x7ffeeff34700
2024-09-19T12:30:34.526547Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x242aa840) invoked
2024-09-19T12:30:34.526828Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 2, 0x182b99b0) invoked
2024-09-19T12:30:34.526992Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x242a74a0) invoked
2024-09-19T12:30:34.527156Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28595fc0) invoked
2024-09-19T12:30:34.527416Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28595ac0) invoked
2024-09-19T12:30:34.527578Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28595e00) invoked
2024-09-19T12:30:34.527769Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28593fc0) invoked
2024-09-19T12:30:34.528266Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28593920) invoked
2024-09-19T12:30:34.528708Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28595aa0) invoked
2024-09-19T12:30:34.529062Z TRACE ThreadId(04) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0xab3cc8c0)), 1, 0xbb89f828) invoked
2024-09-19T12:30:34.573191Z TRACE ThreadId(04) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0xcb7f2220)), 1, 0xbb89e3f0) invoked
2024-09-19T12:30:34.574151Z TRACE ThreadId(04) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 1, 0xbb89f408) invoked
2024-09-19T12:30:34.574500Z TRACE ThreadId(04) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:122: Call IDXGISwapChain::Present trampoline
2024-09-19T12:30:34.575692Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x242a40e0) invoked
2024-09-19T12:30:34.576077Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 2, 0x3c569d90) invoked
2024-09-19T12:30:34.576721Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x242a74a0) invoked
2024-09-19T12:30:34.577193Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x242ad5e0) invoked
2024-09-19T12:30:34.577528Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x242aa840) invoked
2024-09-19T12:30:34.577889Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x242a8c80) invoked
2024-09-19T12:30:34.578237Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28593920) invoked
2024-09-19T12:30:34.578673Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x2859ad80) invoked
2024-09-19T12:30:34.578977Z TRACE ThreadId(03) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x629223e0)), 3, 0x28593fc0) invoked
2024-09-19T12:30:34.579298Z TRACE ThreadId(04) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:146: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0xab3cc8c0)), 1, 0xbb89f828) invoked
2024-09-19T12:30:34.579993Z ERROR ThreadId(04) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:119: Render error: Error { code: HRESULT(0x887A0005), message: "The GPU device instance has been suspended. Use GetDeviceRemovedReason to determine the appropriate action." }
2024-09-19T12:30:34.580299Z TRACE ThreadId(04) hudhook::hooks::dx12: C:\hudhook-0.7.1\src\hooks\dx12.rs:122: Call IDXGISwapChain::Present trampoline

```

**Comment by jaskang** [comment]
Thank you. 
I am sure that the game uses dx12. I tried adding "GetDeviceRemovedReason " and got the following log:


![9effb232aebaa012dc39ec398a4d541](https://github.com/user-attachments/assets/7645ecbb-1305-4599-874c-7f5c64e8caf2)


```shell
2024-10-07T02:34:27.970940Z TRACE ThreadId(01) black_myth_map: src\lib.rs:247: DllMain()
2024-10-07T02:34:27.973664Z  INFO ThreadId(02) black_myth_map: src\lib.rs:69: new
2024-10-07T02:34:27.973984Z  INFO ThreadId(02) black_myth_map: src\lib.rs:72: base_address: 5368709120
2024-10-07T02:34:27.975065Z DEBUG ThreadId(02) hudhook::hooks: vendor\hudhook\src\hooks\mod.rs:87: WNDCLASSEXW { cbSize: 80, style: WNDCLASS_STYLES(3), cbClsExtra: 0, cbWndExtra: 0, hInstance: HINSTANCE(5368709120), hIcon: HICON(0), hCursor: HCURSOR(0), hbrBackground: HBRUSH(0), lpszMenuName: PCWSTR(0x0), lpszClassName: PCWSTR(0x7ff8a10fd4b8), hIconSm: HICON(0) }
2024-10-07T02:34:27.980901Z DEBUG ThreadId(02) hudhook::hooks: vendor\hudhook\src\hooks\mod.rs:107: HWND(329360)
2024-10-07T02:34:28.022564Z TRACE ThreadId(02) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:261: IDXGISwapChain::Present = 0x7ff924544700
2024-10-07T02:34:28.058330Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x187d883c0) invoked
2024-10-07T02:34:28.058559Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 2, 0xc0b5b70) invoked
2024-10-07T02:34:28.058716Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x187d81ee0) invoked
2024-10-07T02:34:28.058872Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x187d8c940) invoked
2024-10-07T02:34:28.059213Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x187d881e0) invoked
2024-10-07T02:34:28.059354Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x187d8fbc0) invoked
2024-10-07T02:34:28.059556Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xd251a140) invoked
2024-10-07T02:34:28.059690Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x272365a0) invoked
2024-10-07T02:34:28.059807Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x132fcc2c0) invoked
2024-10-07T02:34:28.060114Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x12b20d160) invoked
2024-10-07T02:34:28.060798Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x12b20ff40) invoked
2024-10-07T02:34:28.061370Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0xaaa37a10)), 1, 0xba0cf828) invoked
2024-10-07T02:34:28.158831Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:28.483442Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:28.840983Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:29.205632Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:29.582961Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:29.947050Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:30.456981Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:30.821095Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:31.190735Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:31.570420Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:32.095921Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:32.497337Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:32.881917Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:33.428490Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:33.807461Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:34.200394Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:34.585443Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:35.100680Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:35.588660Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:36.101234Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:36.476999Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce1b0) invoked
2024-10-07T02:34:36.517181Z  INFO ThreadId(04) black_myth_map: src\lib.rs:188: 初始化完成
2024-10-07T02:34:36.519921Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6c5fd20)), 1, 0xba0ce3f0) invoked
2024-10-07T02:34:36.524018Z  INFO ThreadId(04) black_myth_map: src\lib.rs:192: render
2024-10-07T02:34:36.525179Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 1, 0xba0cf3d8) invoked
2024-10-07T02:34:36.527200Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:124: Call IDXGISwapChain::Present trampoline
2024-10-07T02:34:36.528288Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x12b20cec0) invoked
2024-10-07T02:34:36.530756Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 2, 0x12b622cb0) invoked
2024-10-07T02:34:36.553171Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x130257d60) invoked
2024-10-07T02:34:36.553598Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0x13025e980) invoked
2024-10-07T02:34:36.553965Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xd25950c0) invoked
2024-10-07T02:34:36.554281Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xd2597dc0) invoked
2024-10-07T02:34:36.554619Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xd2596440) invoked
2024-10-07T02:34:36.554965Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xc6b82f40) invoked
2024-10-07T02:34:36.555344Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xc6b82d20) invoked
2024-10-07T02:34:36.555693Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xc6b85180) invoked
2024-10-07T02:34:36.556010Z TRACE ThreadId(03) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0x6bf6aa0)), 3, 0xc6b83180) invoked
2024-10-07T02:34:36.556349Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:148: ID3D12CommandQueue::ExecuteCommandLists(ID3D12CommandQueue(IUnknown(0xaaa37a10)), 1, 0xba0cf828) invoked
2024-10-07T02:34:36.556643Z  INFO ThreadId(04) black_myth_map: src\lib.rs:192: render
2024-10-07T02:34:36.558096Z ERROR ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:121: Render error: Error { code: HRESULT(0x887A0005), message: "The GPU device instance has been suspended. Use GetDeviceRemovedReason to determine the appropriate action." }, Device Removed Reason: Err(Error { code: HRESULT(0x887A002B), message: "The application is trying to use a resource to which it does not have the required access privileges.\r\nThis is most commonly caused by writing to a shared resource with read-only access." })
2024-10-07T02:34:36.558503Z TRACE ThreadId(04) hudhook::hooks::dx12: vendor\hudhook\src\hooks\dx12.rs:124: Call IDXGISwapChain::Present trampoline

```

**Comment by jaskang** [comment]
If comment "load_texture", it will work.

**Comment by veeenu** [comment]
Without seeing your code, I don't have a lot to work with, unfortunately. I can only offer random guesses.

It looks like the error happens at the second render pass (if that's what the `src/lib.rs:192: render` log means) and you mention `load_texture` so maybe you are loading the same texture multiple times unintentionally?

**Comment by jaskang** [comment]
I'm sorry I provided the wrong information, load_texture is not the key issue. The game crash problem still exists, and I am using basic sample code.

Even when I tried multiple other repositories for hooking DirectX12, including C++, all of them caused the game to crash without exception. Other games work fine though, such as Elden Ring (no more games available for testing).

However, I found some clues:
https://www.unknowncheats.me/forum/unreal-engine-4-a/635608-d3d12-hook-crashes-ue5-03-a.html

**Comment by veeenu** [comment]
That's very interesting. And also very testable - I was not aware that Black Myth used Unreal so could just build a toy game with Unreal and test against that.

I kinda disagree with the solution though. I don't want to go rummaging in the object's vtable.

An alternative solution would be to instead call `ID3D12CommandQueue::GetDesc` in the `ExecuteCommandLists` call, and try to infer which is the correct queue to use.

**Comment by jaskang** [comment]
To sync the progress of this issue, I am a beginner in Rust and making a floating mini-map for this game.

Following your advice, I tried some methods:

Changed `COMMAND_QUEUE` from `OnceCell` to mutable, updated it in `ExecuteCommandLists`. (Honestly, I don't know how to filter it)

![bcdc0a2c52119543d860eeac012306b](https://github.com/user-attachments/assets/edf3ac9f-21b5-42ee-ba57-807084d6410a)

![7d572941d180ac70f6ae29250f9e24e](https://github.com/user-attachments/assets/dc383100-b939-4ddb-b2e7-79ee8f1acb25)

![d81dab3d94acb1f4a86e21ba0218553](https://github.com/user-attachments/assets/3701ba28-5c67-421e-9a0b-de13e37d6809)

Although the code is rough, miraculously the problem was solved like that.

Sincerely thank you.

**Comment by veeenu** [comment]
That's awesome! Would you kindly open a pull request with the changes? That way I can see what the changes are more clearly. Global mutability is something to be handled carefully but I think I would like to upstream some version of this solution.

**Comment by jaskang** [comment]
It is worth mentioning that I found a good example, which obtains the command_queue through swap_chain + offset.

https://github.com/praydog/UEVR/blob/1519726ca22815360d9b024c8b9c53b6296ad06b/src/hooks/D3D12Hook.cpp#L449

![image](https://github.com/user-attachments/assets/6a123e73-f502-492d-92fd-5f0422301553)

**Comment by veeenu** [comment]
> It is worth mentioning that I found a good example, which obtains the `command_queue` through `swap_chain` + `offset`.
> 
> https://github.com/praydog/UEVR/blob/1519726ca22815360d9b024c8b9c53b6296ad06b/src/hooks/D3D12Hook.cpp#L449

That method is actually brilliant. People used to do this by using a single, hardcoded offset, which was very brittle because the offset might change from one version of DirectX 12 to the other, and we have no static guarantee about that. The specific implementation I'm not super confident about, but looping through the vtable and doing _some_ check to ensure we're at a `ID3D12CommandQueue` pointer is a great idea. Not to mention it would allow us to remove one hook and vastly simplify the hook workflow.

I'll do a couple experiment, open a PR and ping you for review!

**Comment by jaskang** [comment]
Great, I am currently using this method and it is working fine, but my code is not robust enough. It couldn't be better if it could be solved directly in hudhook.

**Comment by veeenu** [comment]
@jaskang in [this branch](https://github.com/veeenu/hudhook/compare/dx12-command-queue-vtable?expand=1) I have made an attempt of implementing that strategy. It seems to work in tests, but I don't have DirectX 12 games on my current machine, I'll test it against Elden Ring as soon as I get the chance. I hope it works but I'm not sure it will.

We're using a IDXGISwapChain3 in hudhook and I'm not sure how this code will behave with e.g. IDXGISwapChain2: will the offset at which the command queue pointer is stored be consistent across versions? Do we have static guarantees about those things? If it works with a handful of Dx12 games then we might be good.

**Comment by veeenu** [comment]
I have added a better heuristic. Now the swapchain is recorded first try, and whenever `ID3D12CommandQueue::ExecuteCommandList` is invoked, this command queue is checked against the swapchain to see that it belongs to it. If that's the case, the hook is initialized and no further checks are run against the swap chain and the command queue in their hooked methods. We will still pay the cost of locking the mutex, though we might be able to sidestep that by using a static atomic boolean. This is probably the most robust solution against the issue I raised in my previous comment. Checking the pointers is still a potentially imperfect heuristic, but I think it should always work in practice.

I still need to test it against a UE game, on Windows, and on Elden Ring to check for regressions, but @jaskang I'd like to ask you if you could please confirm it works on Black Myth. If that's the case I'll do my checks, open the PR, merge that in and release 0.8, otherwise we can keep iterating ideas.

**Comment by jaskang** [comment]
![7344fa9b407e2da0d739a3bb65b6f59](https://github.com/user-attachments/assets/374535cb-91f7-4674-a78a-e0efbd7174d1)
![0f2b4395fbee3df153ced265cc5dfca](https://github.com/user-attachments/assets/e7dcfcf1-a5ff-41e1-992f-ba9c71ae9f2f)
![6cebc91dcaf0e8aac5e7c64027ecd23](https://github.com/user-attachments/assets/d2ae283b-4c7d-4375-b854-31d6a1ff646d)

I have tested the Black Myth 、Elden Ring and Hades II. it works very well. 

By the way, may I ask when the many-textures branch can be merged? I have been developing based on this branch and can provide some test samples.

**Comment by veeenu** [comment]
This is awesome, thank you so much!

The textures fix branch was going to get merged as soon as I could run in-depth testing on my own use cases, but you are already successfully working based on it, that gives me more confidence. If I find no issues, I plan to merge both this new branch and that older PR and cut an official release within this week.

**Comment by veeenu** [comment]
@jaskang I just merged #200, then #210 into #209 and rebased #209. I'm now gonna test that against my use cases on Linux. If you want to give the branch another spin and confirm it all works well, we can merge this in and release 0.8.

**Comment by jaskang** [comment]
It works very well. 👍
